python/tchannel/tornado/hyperbahn.py

```diff
@@ -22,18 +22,89 @@
 
 import json
 import logging
+import random
+import time
 
 import tornado.gen
 import tornado.ioloop
 
-from ..messages.error import ErrorCode
-
-DEFAULT_TTL = 60  # seconds
+from ..errors import AdvertiseError
+from ..errors import ProtocolError
+from ..errors import TimeoutError
+from .response import StatusCode
 
+DEFAULT_EXPO_BASE = 1.4  # try this first
+DEFAULT_MAX_DELAY = 10  # sec
+DEFAULT_MAX_ATTEMPT = 7  # pow(1.4, 8) > 10
+DEFAULT_DELAY = 60 * 1000  # ms delay time for successful advertise
+DEFAULT_FIRST_ADVERTISE_TIME = 30  # sec
 
 log = logging.getLogger('tchannel')
 
 
+def _prepare_next_ad(attempt_counter):
+    if attempt_counter:
+        delay_time = random.uniform(
+            0, min(DEFAULT_MAX_DELAY, DEFAULT_EXPO_BASE ** attempt_counter)
+        )
+    else:
+        delay_time = DEFAULT_DELAY
+    attempt_counter = min(attempt_counter, DEFAULT_MAX_ATTEMPT)
+    return (attempt_counter, delay_time)
+
+
+@tornado.gen.coroutine
+def _advertise(tchannel, service):
+    response = None
+    try:
+        response = yield tchannel.request(service='hyperbahn').send(
+            arg1='ad',  # advertise
+            arg2='',
+            arg3=json.dumps({
+                'services': [
+                    {
+                        'serviceName': service,
+                        'cost': 0,
+                    }
+                ]
+            }),
+            headers={'as': 'json'},
+            attempt_times=1,
+        )
+    except (ProtocolError, TimeoutError) as e:
+        log.error('Failed to register with Hyperbahn: %s', e)
+    else:
+        if response.code != StatusCode.ok:
+            log.error('Failed to register with Hyperbahn: %s', response)
+        else:
+            log.info('Successfully register with Hyperbahn')
+
+    raise tornado.gen.Return(response)
+
+
+@tornado.gen.coroutine
+def _exponential_backoff_advertise(tchannel, service, timeout=None):
+    # first advertise rule apply here.
+    attempt_counter = 0
+    start = time.time()
+    while True:
+        if timeout and time.time() - start > 30000:
+            raise AdvertiseError("Failed to register with Hyperbahn.")
+
+        response = yield _advertise(
+            tchannel,
+            service,
+        )
+        if response is not None and response.code is StatusCode.ok:
+            break
+        attempt_counter += 1
+        (attempt_counter, delay_time) = _prepare_next_ad(attempt_counter)
```

Comment:
I see because you have `DEFAULT_MAX_ATTEMPT`. Let's remove the `DEFAULT_DELAY` from `_prepare_next_ad` and we'll be good to go then.